tantar/src/main/java/org/opencadc/tantar/BucketValidator.java

```diff
@@ -611,6 +617,10 @@ public void updateArtifact(final Artifact artifact, final StorageMetadata storag
             @Override
             public boolean hasNext() {
                 if (storageMetadataIterator.hasNext()) {
+                    storageMetadata = storageMetadataIterator.next();
```

Comment:
Calling code can call hasNext() multiple times for no good reason, so you cannot put the "advance" (call to storageMetadataIterator.next()) in here. I usually implement an advance() method that sets the "next item" field, so hasNext() is simply 
```
return storageMetadata != null;
```
Then you have to call advance() in the ctor and in next(), eg
```
public StorageMetadata next() {
   StorageMetadata ret = storageMetadata;
   advance();
   return ret;
}
```
And then advance() does the checking and skipping but only to find the first valid return object (ctor) and when caller actually advances (next).

Reply:
Now I remember using an advance() method in an iterator. Moved the MetadataStorage iterator method to a class, so advance() can be called in the class constructor. And a convenience method to get the class that can be overridden in the unit tests.